libs/ml/include/ml/ops/variable.hpp

```diff
@@ -147,8 +154,11 @@ class Variable : public DataHolder<T>, public Trainable<T>
 
   void ApplyGradient(TensorType const &grad) override
   {
-    ApplyRegularisation();
-    this->data_->InlineAdd(grad);
+    if (!this->value_frozen_)
+    {
+      ApplyRegularisation();
+      this->data_->InlineAdd(grad);
+    }
     ResetGradients();
```

Comment:
I think that ResetGradients should be inside the if-statement? If a variable has gradients from before it was frozen and then it is frozen and then ApplyGradients is called then the gradients from before the freeze will be deleted. I think ResetGradients should only be called if the gradients were applied?